spec/Spec/BlsHelpers.cry

```diff
@@ -53,6 +54,14 @@ compute_powers : {n} (fin n, width n <= 64, n >= 2) => BlsFieldElement -> [n]Bls
 compute_powers x = powers where
         powers = [x] # [(powers@(i-1) * x) % BLS_MODULUS | i <- [1 .. n-1]]
 
+/**
+ * Performs a multi-scalar multiplication between G1 `points` and `scalars`.
+ * @see https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md#multi_exp
+ */
+g1_multi_exp : {len} (fin len, len >= 1) => [len]G1Point -> [len]UInt64 -> G1Point
+g1_multi_exp points scalars = results!0 where
+    results = [G1_INFINITY] # [g1_add (results@i) (g1_multi (points@i) (scalars@i)) | i <- [0 .. len-1]]
```

Comment:
One has to "generate" the Python spec and then the function body is revealed. Maybe I should add that as a comment?

Reply:
I thought the exact same thing.

Reply:
@marsella Here's the implementation:
```
def multi_exp(points, integers):
    """
    Performs a multi-scalar multiplication between
    `points` and `integers`.
    `points` can either be in G1 or G2.
    """
    # Since this method accepts either G1 or G2, we need to know
    # the type of the point to return. Hence, we need at least one point.
    if not points or not integers:
        raise Exception("Cannot call multi_exp with zero points or zero integers")

    if bls == arkworks_bls or bls == fastest_bls:
        # Convert integers into arkworks Scalars
        scalars = []
        for integer in integers:
            int_as_bytes = integer.to_bytes(32, 'little')
            scalars.append(arkworks_Scalar.from_le_bytes(int_as_bytes))

        # Check if we need to perform a G1 or G2 multiexp
        if isinstance(points[0], arkworks_G1):
            return arkworks_G1.multiexp_unchecked(points, scalars)
        elif isinstance(points[0], arkworks_G2):
            return arkworks_G2.multiexp_unchecked(points, scalars)
        else:
            raise Exception("Invalid point type")

    result = None
    if isinstance(points[0][0], FQ):
        result = Z1()
    elif isinstance(points[0][0], FQ2):
        result = Z2()
    else:
        raise Exception("Invalid point type")

    for point, scalar in zip(points, integers):
        result = add(result, multiply(point, scalar))
    return result
 ```

Reply:
First difference is that they are handling both G1 and G2 multiplication, mine is only handling G1 for now.
Second difference is that it shortcuts the whole computation when the first point is infinity.

Reply:
I found the link: https://github.com/ethereum/consensus-specs/blob/85e2452301382a9d099df16f38deac8178355358/tests/core/pyspec/eth2spec/utils/bls.py#L262

I'll add it to the documentation.

Reply:
Thank you!